File: mypy/astwire.py
"""Generic AST node serialization for the type_kernel strangler-fig seam.

Serializes AST nodes (mypy.nodes.Node subclasses) into a binary wire format
that Rust can traverse. This is a generic serializer: it introspects
``__slots__`` and serializes any field that is a ``Node`` (recursively) or a
list/tuple of ``Node`` objects. Scalar fields (line, column, name, kind, etc.)
are not serialized — the format captures tree structure and node kind only,
which is sufficient for the traverser seekers (has_return_statement,
has_yield_expression, etc.) that only need to walk the tree and match on
node kind.

Wire format per node::

    tag:u8                          # node kind tag (from nodes.py constants)
    num_child_fields:varint         # count of serialized child fields
    child_field*                    # each child field:
        LITERAL_NONE                 # if the field is None
        | node                       # if the field is a single Node
        | LIST_GEN size item*        # if the field is a list of Nodes

Leaf nodes (IntExpr, StrExpr, etc.) have no child fields. The tag alone
identifies them.
"""

from __future__ import annotations

from typing import Any

from librt.internal import write_int as write_int_bare

from mypy import nodes
from mypy.cache import END_TAG, LIST_GEN, LITERAL_NONE, WriteBuffer, write_tag

# Build the tag lookup: class name -> tag value.
# Maps from the Python class to its wire tag constant.
_NODE_TAGS: dict[type, int] = {}


# Wire-format-only tags for expression types that have no Final[Tag]
# in nodes.py. Range 150-159 is unused by the cache format (reserved:

# 50-79 symbols, 80-149 types). These are local to the astwire
# format and never written to/read from the mypy metadata cache.
ASTWIRE_CAST_EXPR: int = 150
ASTWIRE_ASSERT_TYPE_EXPR: int = 151
ASTWIRE_REVEAL_EXPR: int = 152
ASTWIRE_SUPER_EXPR: int = 153
ASTWIRE_TYPE_APPLICATION: int = 154
ASTWIRE_TYPE_ALIAS_EXPR: int = 155
ASTWIRE_NAMEDTUPLE_EXPR: int = 156
ASTWIRE_TYPEDDICT_EXPR: int = 157
ASTWIRE_ENUM_CALL_EXPR: int = 158
ASTWIRE_PROMOTE_EXPR: int = 159


def _register_node_tags() -> None:
    """Populate ``_NODE_TAGS`` from mypy.nodes tag constants."""
    tag_map = {
        nodes.MYPY_FILE: nodes.MypyFile,
        nodes.OVERLOADED_FUNC_DEF: nodes.OverloadedFuncDef,
        nodes.FUNC_DEF: nodes.FuncDef,
        nodes.DECORATOR: nodes.Decorator,
        nodes.VAR: nodes.Var,
        nodes.TYPE_VAR_EXPR: nodes.TypeVarExpr,
        nodes.PARAM_SPEC_EXPR: nodes.ParamSpecExpr,
        nodes.TYPE_VAR_TUPLE_EXPR: nodes.TypeVarTupleExpr,
        nodes.TYPE_INFO: nodes.TypeInfo,
        nodes.TYPE_ALIAS: nodes.TypeAlias,
        nodes.CLASS_DEF: nodes.ClassDef,
        nodes.SYMBOL_TABLE_NODE: nodes.SymbolTableNode,
        nodes.EXPR_STMT: nodes.ExpressionStmt,
        nodes.CALL_EXPR: nodes.CallExpr,
        nodes.NAME_EXPR: nodes.NameExpr,
        nodes.STR_EXPR: nodes.StrExpr,
        nodes.IMPORT: nodes.Import,
        nodes.MEMBER_EXPR: nodes.MemberExpr,
        nodes.OP_EXPR: nodes.OpExpr,
        nodes.INT_EXPR: nodes.IntExpr,
        nodes.IF_STMT: nodes.IfStmt,
        nodes.ASSIGNMENT_STMT: nodes.AssignmentStmt,
        nodes.TUPLE_EXPR: nodes.TupleExpr,
        nodes.BLOCK: nodes.Block,
        nodes.INDEX_EXPR: nodes.IndexExpr,
        nodes.LIST_EXPR: nodes.ListExpr,
        nodes.SET_EXPR: nodes.SetExpr,
        nodes.RETURN_STMT: nodes.ReturnStmt,
        nodes.WHILE_STMT: nodes.WhileStmt,
        nodes.COMPARISON_EXPR: nodes.ComparisonExpr,
        nodes.BOOL_OP_EXPR: nodes.OpExpr,
        nodes.FUNC_DEF_STMT: nodes.FuncDef,
        nodes.PASS_STMT: nodes.PassStmt,
        nodes.FLOAT_EXPR: nodes.FloatExpr,
        nodes.UNARY_EXPR: nodes.UnaryExpr,
        nodes.DICT_EXPR: nodes.DictExpr,
        nodes.COMPLEX_EXPR: nodes.ComplexExpr,
        nodes.SLICE_EXPR: nodes.SliceExpr,
        nodes.TEMP_NODE: nodes.TempNode,
        nodes.RAISE_STMT: nodes.RaiseStmt,
        nodes.BREAK_STMT: nodes.BreakStmt,
        nodes.CONTINUE_STMT: nodes.ContinueStmt,
        nodes.GENERATOR_EXPR: nodes.GeneratorExpr,
        nodes.YIELD_EXPR: nodes.YieldExpr,
        nodes.YIELD_FROM_EXPR: nodes.YieldFromExpr,
        nodes.LIST_COMPREHENSION: nodes.ListComprehension,
        nodes.SET_COMPREHENSION: nodes.SetComprehension,
        nodes.DICT_COMPREHENSION: nodes.DictionaryComprehension,
        nodes.IMPORT_FROM: nodes.ImportFrom,
        nodes.ASSERT_STMT: nodes.AssertStmt,
        nodes.FOR_STMT: nodes.ForStmt,
        nodes.WITH_STMT: nodes.WithStmt,
        nodes.OPERATOR_ASSIGNMENT_STMT: nodes.OperatorAssignmentStmt,
        nodes.TRY_STMT: nodes.TryStmt,
        nodes.ELLIPSIS_EXPR: nodes.EllipsisExpr,
        nodes.CONDITIONAL_EXPR: nodes.ConditionalExpr,
        nodes.DEL_STMT: nodes.DelStmt,
        nodes.LAMBDA_EXPR: nodes.LambdaExpr,
        nodes.ASSIGNMENT_EXPR: nodes.AssignmentExpr,
        nodes.STAR_EXPR: nodes.StarExpr,
        nodes.BYTES_EXPR: nodes.BytesExpr,
        nodes.GLOBAL_DECL: nodes.GlobalDecl,
        nodes.NONLOCAL_DECL: nodes.NonlocalDecl,
        nodes.AWAIT_EXPR: nodes.AwaitExpr,
        nodes.IMPORT_ALL: nodes.ImportAll,
        nodes.MATCH_STMT: nodes.MatchStmt,
        nodes.TYPE_ALIAS_STMT: nodes.TypeAliasStmt,
        nodes.TSTRING_EXPR: nodes.TemplateStrExpr,
        # Expression types without their own Final[Tag] — assigned
        # wire-format-only tags so the Rust traverser can see into them.
        ASTWIRE_CAST_EXPR: nodes.CastExpr,
        ASTWIRE_ASSERT_TYPE_EXPR: nodes.AssertTypeExpr,
        ASTWIRE_REVEAL_EXPR: nodes.RevealExpr,
        ASTWIRE_SUPER_EXPR: nodes.SuperExpr,
        ASTWIRE_TYPE_APPLICATION: nodes.TypeApplication,
        ASTWIRE_TYPE_ALIAS_EXPR: nodes.TypeAliasExpr,
        ASTWIRE_NAMEDTUPLE_EXPR: nodes.NamedTupleExpr,
        ASTWIRE_TYPEDDICT_EXPR: nodes.TypedDictExpr,
        ASTWIRE_ENUM_CALL_EXPR: nodes.EnumCallExpr,
        ASTWIRE_PROMOTE_EXPR: nodes.PromoteExpr,
    }
    # Patterns.
    from mypy import patterns as p

    tag_map.update(
        {
            nodes.AS_PATTERN: p.AsPattern,
            nodes.OR_PATTERN: p.OrPattern,
            nodes.VALUE_PATTERN: p.ValuePattern,
            nodes.SINGLETON_PATTERN: p.SingletonPattern,
            nodes.SEQUENCE_PATTERN: p.SequencePattern,
            nodes.STARRED_PATTERN: p.StarredPattern,
            nodes.MAPPING_PATTERN: p.MappingPattern,
            nodes.CLASS_PATTERN: p.ClassPattern,
        }
    )
    for tag, cls in tag_map.items():
        _NODE_TAGS[cls] = tag


_register_node_tags()


# Slots are fixed per class; cache the per-class slot lists once. AST
# node classes are created before this module is imported, so the cache
# cannot go stale during a run.
_SLOT_CACHE: dict[type, list[str]] = {}


def _get_all_slots(cls: type) -> list[str]:
    cached = _SLOT_CACHE.get(cls)
    if cached is not None:
        return cached
    result: list[str] = []
    skip = {object, nodes.Context, nodes.Node, nodes.Statement, nodes.Expression}
    try:
        from mypy.patterns import Pattern

        skip.add(Pattern)
    except ImportError:
        pass
    for base in cls.__mro__:
        if base in skip:
            continue
        slots = getattr(base, "__slots__", ())
        if isinstance(slots, str):
            result.append(slots)
        else:
            result.extend(slots)
    # Deduplicate while preserving order.
    seen: set[str] = set()
    out: list[str] = []
    for s in result:
        if s not in seen:
            seen.add(s)
            out.append(s)
    _SLOT_CACHE[cls] = out
    return out


def serialize_node(node: nodes.Node | None, buf: WriteBuffer) -> None:
    """Serialize a node into ``buf`` as tag + child fields (iterative).

    Uses an explicit task stack instead of recursion so that deeply nested
    ASTs don't hit Python's recursion limit.

    Lists-of-lists (e.g. GeneratorExpr.condlists) serialize as a
    LIST_GEN whose items are themselves LIST_GEN blocks, so Rust sees
    the full nested structure.
    """
    if node is None:
        write_tag(buf, LITERAL_NONE)
        return

    # Track visited nodes to break cycles (e.g. NameExpr.node → FuncDef
    # → body → NameExpr). Re-serialized nodes write LITERAL_NONE.
    visited: set[int] = set()
    # Task stack: ("end", None), ("none", None), ("list", items), ("node", node)
    stack: list[tuple[str, Any]] = [("node", node)]
    while stack:
        task = stack.pop()
        kind = task[0]
        if kind == "end":
            write_tag(buf, END_TAG)
        elif kind == "none":
            write_tag(buf, LITERAL_NONE)
        elif kind == "list":
            items = task[1]
            # Nested list: emit LIST_GEN with each item a LIST_GEN block.
            if items and isinstance(items[0], (list, tuple)):
                write_tag(buf, LIST_GEN)
                write_int_bare(buf, len(items))
                for item in reversed(items):
                    stack.append(("list", item))
                continue
            write_tag(buf, LIST_GEN)
            write_int_bare(buf, len(items))
            for item in reversed(items):
                if isinstance(item, nodes.Node):
                    stack.append(("node", item))
                else:
                    stack.append(("none", None))
        elif kind == "node":
            n = task[1]
            cls = type(n)
            tag = _NODE_TAGS.get(cls)
            if tag is None or isinstance(n, nodes.TypeInfo):
                write_tag(buf, LITERAL_NONE)
                continue
            # Break cycles: skip already-visited nodes.
            if id(n) in visited:
                write_tag(buf, LITERAL_NONE)
                continue
            visited.add(id(n))
            write_tag(buf, tag)
            slots = _get_all_slots(cls)

            # Collect child fields (Node or list-of-Node, or nested list).
            # Lists are appended by reference: the serializer only reads
            # them (reversed() iteration in the "list" task), never mutates.
            child_fields: list[Any] = []
            for slot in slots:
                value = getattr(n, slot, None)
                if isinstance(value, nodes.Node):
                    child_fields.append(value)
                elif isinstance(value, (list, tuple)):
                    # Keep list fields (Nodes/None, or nested lists of
                    # them). Dispatch on value[0] so flat lists
                    # skip the nested-row check.
                    if not value:
                        child_fields.append(None)
                    else:
                        first = value[0]
                        if isinstance(first, nodes.Node) or first is None:
                            if all((isinstance(i, nodes.Node) or i is None) for i in value):
                                child_fields.append(value)
                            else:
                                child_fields.append(None)
                        elif isinstance(first, (list, tuple)):
                            ok = True
                            for row in value:
                                if not isinstance(row, (list, tuple)):
                                    ok = False
                                    break
                                for i in row:
                                    if not (isinstance(i, nodes.Node) or i is None):
                                        ok = False
                                        break
                                if not ok:
                                    break
                            child_fields.append(value if ok else None)
                        else:
                            child_fields.append(None)
                else:
                    child_fields.append(None)

            write_int_bare(buf, len(child_fields))
            # Push END_TAG first (LIFO → processed last), then fields
            # in reverse order so they pop in forward order.
            stack.append(("end", None))
            for field in reversed(child_fields):
                if field is None:
                    stack.append(("none", None))
                elif isinstance(field, nodes.Node):
                    stack.append(("node", field))
                else:
                    stack.append(("list", field))

File: crates/type_kernel/src/astwire.rs
#![allow(dead_code)]

//! AST node wire format: Rust-side reader for the generic node serializer
//! (mirrors `mypy/astwire.py`). Provides a tree structure with variant tag
//! and child fields, sufficient for traversal-only visitors (the traverser
//! seekers/collectors that match on node kind and walk children).
//!
//! Wire format (see astwire.py):
//!   LITERAL_NONE (2)          → None node
//!   tag:u8 count:varint child* END_TAG
//!     child = LITERAL_NONE | node | LIST_GEN size item*
//!
//! This module does NOT parse scalar fields (line, column, name, kind) — the
//! Python serializer skips them. Only structure + node kind survives the
//! round-trip, which is what the seekers need.

use crate::wire::{read_int_bare, read_tag, ReadBuffer, WireError, WriteBuffer};

// Shared wire tags (mirror mypy.cache).
const LITERAL_NONE: u8 = 2;
const LIST_GEN: u8 = 20;
const END_TAG: u8 = 255;

// ---------------------------------------------------------------------------
// Node tags (mirror mypy/nodes.py Final[Tag] constants)
// ---------------------------------------------------------------------------

pub(crate) const MYPY_FILE: u8 = 50;
pub(crate) const OVERLOADED_FUNC_DEF: u8 = 51;
pub(crate) const FUNC_DEF: u8 = 52;
pub(crate) const DECORATOR: u8 = 53;
pub(crate) const VAR: u8 = 54;
pub(crate) const CLASS_DEF: u8 = 60;
pub(crate) const EXPR_STMT: u8 = 160;
pub(crate) const CALL_EXPR: u8 = 161;
pub(crate) const NAME_EXPR: u8 = 162;
pub(crate) const STR_EXPR: u8 = 163;
pub(crate) const IMPORT: u8 = 164;
pub(crate) const MEMBER_EXPR: u8 = 165;
pub(crate) const OP_EXPR: u8 = 166;
pub(crate) const INT_EXPR: u8 = 167;
pub(crate) const IF_STMT: u8 = 168;
pub(crate) const ASSIGNMENT_STMT: u8 = 169;
pub(crate) const TUPLE_EXPR: u8 = 170;
pub(crate) const BLOCK: u8 = 171;
pub(crate) const INDEX_EXPR: u8 = 172;
pub(crate) const LIST_EXPR: u8 = 173;
pub(crate) const SET_EXPR: u8 = 174;
pub(crate) const RETURN_STMT: u8 = 175;
pub(crate) const WHILE_STMT: u8 = 176;
pub(crate) const COMPARISON_EXPR: u8 = 177;
pub(crate) const FUNC_DEF_STMT: u8 = 179;
pub(crate) const PASS_STMT: u8 = 180;
pub(crate) const FLOAT_EXPR: u8 = 181;
pub(crate) const UNARY_EXPR: u8 = 182;
pub(crate) const DICT_EXPR: u8 = 183;
pub(crate) const COMPLEX_EXPR: u8 = 184;
pub(crate) const SLICE_EXPR: u8 = 185;
pub(crate) const TEMP_NODE: u8 = 186;
pub(crate) const RAISE_STMT: u8 = 187;
pub(crate) const BREAK_STMT: u8 = 188;
pub(crate) const CONTINUE_STMT: u8 = 189;
pub(crate) const GENERATOR_EXPR: u8 = 190;
pub(crate) const YIELD_EXPR: u8 = 191;
pub(crate) const YIELD_FROM_EXPR: u8 = 192;
pub(crate) const LIST_COMPREHENSION: u8 = 193;
pub(crate) const SET_COMPREHENSION: u8 = 194;
pub(crate) const DICT_COMPREHENSION: u8 = 195;
pub(crate) const IMPORT_FROM: u8 = 196;
pub(crate) const ASSERT_STMT: u8 = 197;
pub(crate) const FOR_STMT: u8 = 198;
pub(crate) const WITH_STMT: u8 = 199;
pub(crate) const OPERATOR_ASSIGNMENT_STMT: u8 = 200;
pub(crate) const TRY_STMT: u8 = 201;
pub(crate) const ELLIPSIS_EXPR: u8 = 202;
pub(crate) const CONDITIONAL_EXPR: u8 = 203;
pub(crate) const DEL_STMT: u8 = 204;
pub(crate) const LAMBDA_EXPR: u8 = 207;
pub(crate) const ASSIGNMENT_EXPR: u8 = 208;
pub(crate) const STAR_EXPR: u8 = 209;
pub(crate) const BYTES_EXPR: u8 = 210;
pub(crate) const GLOBAL_DECL: u8 = 211;
pub(crate) const NONLOCAL_DECL: u8 = 212;
pub(crate) const AWAIT_EXPR: u8 = 213;
pub(crate) const IMPORT_ALL: u8 = 215;
pub(crate) const MATCH_STMT: u8 = 216;
pub(crate) const AS_PATTERN: u8 = 217;
pub(crate) const OR_PATTERN: u8 = 218;
pub(crate) const VALUE_PATTERN: u8 = 219;
pub(crate) const SINGLETON_PATTERN: u8 = 220;
pub(crate) const SEQUENCE_PATTERN: u8 = 221;
pub(crate) const STARRED_PATTERN: u8 = 222;
pub(crate) const MAPPING_PATTERN: u8 = 223;
pub(crate) const CLASS_PATTERN: u8 = 224;
pub(crate) const TYPE_ALIAS_STMT: u8 = 225;

// ---------------------------------------------------------------------------
// Node tree representation
// ---------------------------------------------------------------------------

/// A child field: either None, a single node, or a list of nodes.
#[derive(Debug, Clone)]
pub(crate) enum ChildField {
    None,
    Node(AstNode),
    List(Vec<AstNode>),
}

/// An AST node: variant tag + child fields. Structure only (no scalar data).
#[derive(Debug, Clone)]
pub(crate) struct AstNode {
    pub tag: u8,
    pub children: Vec<ChildField>,
}

impl AstNode {
    /// Yield all direct child nodes (flattening list fields).
    pub fn child_nodes(&self) -> Vec<&AstNode> {
        let mut out = Vec::new();
        for field in &self.children {
            match field {
                ChildField::Node(n) => out.push(n),
                ChildField::List(items) => out.extend(items.iter()),
                ChildField::None => {}
            }
        }
        out
    }
}

// ---------------------------------------------------------------------------
// Reader (mirrors astwire.py serialize_node)
// ---------------------------------------------------------------------------

/// Read a node from the buffer. Returns `None` if the next tag is
/// `LITERAL_NONE` (the Python serializer writes None nodes as LITERAL_NONE).
pub(crate) fn read_node(buf: &mut ReadBuffer<'_>) -> Result<Option<AstNode>, WireError> {
    let tag = read_tag(buf)?;
    if tag == LITERAL_NONE {
        return Ok(None);
    }
    let count = read_int_bare(buf)? as usize;
    let mut children = Vec::with_capacity(count);
    for _ in 0..count {
        let field = read_child_field(buf)?;
        children.push(field);
    }
    let end = read_tag(buf)?;
    if end != END_TAG {
        return Err(WireError::invalid(format!("expected END_TAG, got {end}")));
    }
    Ok(Some(AstNode { tag, children }))
}

fn read_child_field(buf: &mut ReadBuffer<'_>) -> Result<ChildField, WireError> {
    // Peek: if next is a node tag, read it as Node. If LIST_GEN, read list.
    // If LITERAL_NONE, it's None. The serializer writes LITERAL_NONE for
    // None fields, LIST_GEN for lists, or a node tag for single nodes.
    let tag = read_tag(buf)?;
    if tag == LITERAL_NONE {
        return Ok(ChildField::None);
    }
    if tag == LIST_GEN {
        let size = read_int_bare(buf)? as usize;
        let mut items = Vec::with_capacity(size);
        for _ in 0..size {
            if let Some(node) = read_node(buf)? {
                items.push(node);
            }
        }
        return Ok(ChildField::List(items));
    }
    // It's a node tag — re-read the node with this tag already consumed.
    // We need to read count + children + END_TAG.
    let count = read_int_bare(buf)? as usize;
    let mut children = Vec::with_capacity(count);
    for _ in 0..count {
        let field = read_child_field(buf)?;
        children.push(field);
    }
    let end = read_tag(buf)?;
    if end != END_TAG {
        return Err(WireError::invalid(format!("expected END_TAG, got {end}")));
    }
    Ok(ChildField::Node(AstNode { tag, children }))
}

// ---------------------------------------------------------------------------
// Writer (for round-trip tests)
// ---------------------------------------------------------------------------

pub(crate) fn write_node(buf: &mut WriteBuffer, node: &AstNode) {
    buf.push(node.tag);
    crate::wire::write_int_bare(buf, node.children.len() as i64).expect("write child count");
    for field in &node.children {
        write_child_field(buf, field);
    }
    buf.push(END_TAG);
}

fn write_child_field(buf: &mut WriteBuffer, field: &ChildField) {
    match field {
        ChildField::None => buf.push(LITERAL_NONE),
        ChildField::Node(n) => write_node(buf, n),
        ChildField::List(items) => {
            buf.push(LIST_GEN);
            crate::wire::write_int_bare(buf, items.len() as i64).expect("write list size");
            for item in items {
                write_node(buf, item);
            }
        }
    }
}

/// Decode a node from bytes (PyO3 entry point helper).
pub(crate) fn decode_node(bytes: &[u8]) -> Option<AstNode> {
    let mut buf = ReadBuffer::new(bytes);
    read_node(&mut buf).ok().flatten()
}

// ---------------------------------------------------------------------------
// Node kind predicates (for seekers)
// ---------------------------------------------------------------------------

/// Check if a node is a ReturnStmt.
pub(crate) fn is_return_stmt(tag: u8) -> bool {
    tag == RETURN_STMT
}

/// Check if a node is a YieldExpr.
pub(crate) fn is_yield_expr(tag: u8) -> bool {
    tag == YIELD_EXPR
}

/// Check if a node is a YieldFromExpr.
pub(crate) fn is_yield_from_expr(tag: u8) -> bool {
    tag == YIELD_FROM_EXPR
}

/// Check if a node is an AwaitExpr.
pub(crate) fn is_await_expr(tag: u8) -> bool {
    tag == AWAIT_EXPR
}

/// Check if a node is a StrExpr.
pub(crate) fn is_str_expr(tag: u8) -> bool {
    tag == STR_EXPR
}

/// Check if a node is a NameExpr.
pub(crate) fn is_name_expr(tag: u8) -> bool {
    tag == NAME_EXPR
}

/// Check if a node is a MemberExpr.
pub(crate) fn is_member_expr(tag: u8) -> bool {
    tag == MEMBER_EXPR
}

/// Check if a node is a FuncDef (the FUNC_DEF tag, used by both FuncDef
/// and the FUNC_DEF_STMT alias).
pub(crate) fn is_func_def(tag: u8) -> bool {
    tag == FUNC_DEF || tag == FUNC_DEF_STMT || tag == OVERLOADED_FUNC_DEF
}

/// Check if a node is an AssignmentStmt.
pub(crate) fn is_assignment_stmt(tag: u8) -> bool {
    tag == ASSIGNMENT_STMT
}

#[cfg(test)]
mod tests {
    use super::*;

    fn make_return(expr: Option<AstNode>) -> AstNode {
        let children = match expr {
            Some(n) => vec![ChildField::Node(n)],
            None => vec![ChildField::None],
        };
        AstNode {
            tag: RETURN_STMT,
            children,
        }
    }

    fn make_int() -> AstNode {
        AstNode {
            tag: INT_EXPR,
            children: vec![],
        }
    }

    fn make_yield() -> AstNode {
        AstNode {
            tag: YIELD_EXPR,
            children: vec![ChildField::None],
        }
    }

    fn make_block(stmts: Vec<AstNode>) -> AstNode {
        AstNode {
            tag: BLOCK,
            children: vec![ChildField::List(stmts)],
        }
    }

    fn round_trip(node: &AstNode) -> AstNode {
        let mut buf = WriteBuffer::new();
        write_node(&mut buf, node);
        let bytes = buf.into_bytes();
        decode_node(&bytes).expect("decode failed")
    }

    #[test]
    fn test_round_trip_return_with_expr() {
        let node = make_return(Some(make_int()));
        let result = round_trip(&node);
        assert_eq!(result.tag, RETURN_STMT);
        assert_eq!(result.children.len(), 1);
        match &result.children[0] {
            ChildField::Node(n) => assert_eq!(n.tag, INT_EXPR),
            _ => panic!("expected Node child"),
        }
    }

    #[test]
    fn test_round_trip_return_none() {
        let node = make_return(None);
        let result = round_trip(&node);
        assert_eq!(result.tag, RETURN_STMT);
        assert!(matches!(result.children[0], ChildField::None));
    }

    #[test]
    fn test_round_trip_block_with_list() {
        let block = make_block(vec![make_return(Some(make_int())), make_yield()]);
        let result = round_trip(&block);
        assert_eq!(result.tag, BLOCK);
        match &result.children[0] {
            ChildField::List(items) => {
                assert_eq!(items.len(), 2);
                assert_eq!(items[0].tag, RETURN_STMT);
                assert_eq!(items[1].tag, YIELD_EXPR);
            }
            _ => panic!("expected List child"),
        }
    }

    #[test]
    fn test_child_nodes_flattens() {
        let block = make_block(vec![make_return(None), make_yield()]);
        let children = block.child_nodes();
        assert_eq!(children.len(), 2);
    }

    #[test]
    fn test_predicates() {
        assert!(is_return_stmt(RETURN_STMT));
        assert!(!is_return_stmt(YIELD_EXPR));
        assert!(is_yield_expr(YIELD_EXPR));
        assert!(is_yield_from_expr(YIELD_FROM_EXPR));
        assert!(is_await_expr(AWAIT_EXPR));
        assert!(is_str_expr(STR_EXPR));
        assert!(is_name_expr(NAME_EXPR));
        assert!(is_member_expr(MEMBER_EXPR));
        assert!(is_func_def(FUNC_DEF));
        assert!(is_assignment_stmt(ASSIGNMENT_STMT));
    }

    #[test]
    fn test_empty_node() {
        let node = AstNode {
            tag: INT_EXPR,
            children: vec![],
        };
        let result = round_trip(&node);
        assert_eq!(result.tag, INT_EXPR);
        assert!(result.children.is_empty());
        assert!(result.child_nodes().is_empty());
    }
}
